src/Development/IDE/Core/Rules.hs

```diff
@@ -894,7 +894,7 @@ getLinkableType f = do
 
 needsCompilationRule :: Rules ()
 needsCompilationRule = defineEarlyCutoff $ \NeedsCompilation file -> do
-  ((ms,_),_) <- useWithStale_ GetModSummaryWithoutTimestamps file
+  (ms,_) <- use_ GetModSummaryWithoutTimestamps file
```

Comment:
Consider that `NeedsCompilation` is the only case where information flows from the top of the module hierarchy to the bottom i.e. to dependencies from root modules. This means that a failure in a module high up in the hierarchy can cause everything it (transitively) depends on to fail.

In this case, if `NeedsCompilation` fails for something at the top, its result will be under-approximated to be `False` in the dependencies. However, if this module actually used TH,  this will cause a re-evaluation of `GetModIface` for all dependencies (since we don't need to generate object code anymore). Once the module is fixed, we will discover that we actually needed all the object code that we just threw away, and thus have to recompile all dependencies once again, this time keeping the object code.

This is a lot of wasted work, so we need to think carefully about removing it.

Reply:
I agree that removing this `useWithStale_` is more tricky. What can be done:

1. Leave it as it is
2. Move the `useWithStale` to the recursive `NeedsCompilation` call
3. Remove the over approximation 

3 is not an option, it would be very counterintuitive behaviour for diagnostics to propagate from dependencies to root modules.

2 and 1 are semantically close but not equivalent. I think 2 is slightly better: it has fail-fast semantics and it's more intentional. But that said the risk of getting it wrong is high and it's not easy to test it, so I'll leave it as it is